procrastinate/worker.py

```diff
@@ -289,74 +194,154 @@ async def run_job(self, job: jobs.Job, worker_id: int) -> None:
+            ),
         )
 
-        contexts = [
-            context for context in self.current_contexts.values() if context.job
-        ]
-        now = time.time()
-        for context in contexts:
+        job_semaphore = asyncio.Semaphore(self.concurrency)
+
+        running_jobs: dict[JobContext, asyncio.Task] = {}
+        side_tasks = [asyncio.create_task(self.periodic_deferrer())]
+        if self.wait and self.listen_notify:
+            listener_coro = self.app.job_manager.listen_for_jobs(
+                event=notify_event,
+                queues=self.queues,
+            )
+            side_tasks.append(asyncio.create_task(listener_coro, name="listener"))
+
+        try:
+            context = contextlib.nullcontext()
+            if self.install_signal_handlers:
+                context = signals.on_stop(self.stop)
+            with context:
+                """Processes jobs until cancelled or until there is no more available job (wait=False)"""
+                while True:
+                    out_of_job = None
+                    while not out_of_job:
+                        # acquire job_semaphore so that a new job is not fetched if maximum concurrency is reached
+                        async with job_semaphore:
+                            job = await self.app.job_manager.fetch_job(
+                                queues=self.queues
+                            )
+                        if job:
+                            # job_semaphore should be acquired straight because it is
+                            # only acquired when not at full capacity at this time
+                            # however, shield it from cancellation in the unlikely event the worker
+                            # is cancelled at that precise time to not abandon the job
+                            await asyncio.shield(job_semaphore.acquire())
+
+                            job_context = self.base_context.evolve(
+                                additional_context=self.base_context.additional_context.copy(),
+                                job=job,
+                                task=self.app.tasks.get(job.task_name),
+                            )
+                            job_task = asyncio.create_task(
+                                self._process_job(job_context)
+                            )
+                            running_jobs[job_context] = job_task
+
+                            def on_job_complete(
+                                job_context: JobContext, _: asyncio.Task
+                            ):
+                                del running_jobs[job_context]
+                                job_semaphore.release()
+
+                            job_task.add_done_callback(
+                                functools.partial(on_job_complete, job_context)
+                            )
+                        else:
+                            out_of_job = True
```

Comment:
It looks like you could break here and avoid the `out_ob_job` variable completely.

Also, if you invert the `if` and the `else`clause, you could remove one level of indent for all the code above.